### ISSUE TITLE
Handling unconnected gradients in Scipy optimizer

### ISSUE THREAD

**Issue by st--** [GPflow/GPflow#1600]
# Feature request

Our Scipy optimizer wrapper should explicitly handle unconnected gradients.

## Motivation

Currently, the GPflow Scipy optimizer wrapper (`gpflow.optimizers.Scipy`) simply errors (with one of the typically unhelpful TensorFlow error messages) when one of the passed-in `variables` is not connected to the loss to be minimized. This can easily happen, for example when using an SVGP model with a `Constant` kernel (which might be used as a very simple baseline) but having inducing locations not explicitly set to non-trainable. This could also happen erroneously, e.g. when calling `Scipy().minimize(model1.training_loss, model2.trainable_variables)`.

## Proposal

**Describe the solution you would like**
An explicit error message when the variables are not connected to the loss (and hence the `grads` list contains `None` entries).

**What alternatives have you considered?**
We could also simply change the call for the gradient computation to use `unconnected_gradients=tf.UnconnectedGradients.ZERO`. This might silently hide bug cases like `minimize(model1.training_loss, model2.trainable_variables)`, so perhaps this option could be an extra flag to the minimize() call?

**Comment by antonykamp** [comment]
Hey,
If this issue is still active, I would be happy to work on it. 
One question: is checking whether `gradient` returns a zero tensor with `tf.math.count_nonzero` appropriate?

**Comment by st--** [comment]
Hi @antonykamp ! thanks for your interest. Yes, nothing has changed about this issue - if you try to optimize a loss that's not connected to the variables you pass in, you still get an unhelpful error message.
Regarding your question, can you expand a bit more on how you'd want to use count_nonzero?

**Comment by antonykamp** [comment]
> Can you expand a bit more on how you'd want to use count_nonzero?

Of course! When using `unconnected_gradients=tf.UnconnectedGradients.ZERO` and the loss is not connected to the variables, the associated item in the gradient equals a zero tensor. GPflow could test for this by counting the nonzeros of each item. Should the situation arise that every value equals zero, GPflow warns the user, eg.

https://github.com/antonykamp/GPflow/blob/599bc49ed902c2296a550c8f0409dcc9bdcef900/gpflow/optimizers/scipy.py#L181

**Comment by st--** [comment]
Ah! You mean to give a warning, yet continue running?
Will the if check you added in your link work correctly within tf.function()? even if it does, it might make the call a bit slower. How about moving the check to https://github.com/antonykamp/GPflow/blob/599bc49ed902c2296a550c8f0409dcc9bdcef900/gpflow/optimizers/scipy.py#L119 (where we've got to be in numpy-land anyways), and maybe having a flag so it only actually checks on the first call ?

**Comment by antonykamp** [comment]
> Ah! You mean to give a warning, yet continue running?

As you pointed out, sometimes its desired behavior eg. in bug cases. Am I right? :)

Moving the check was a great idea ofc :D I would add the new flag as a simple attribute of `Scipy`-objects (and `uncon_gradients_handle` stays?)

**Comment by antonykamp** [comment]
Should we move this discussion to a PR? :)

**Comment by st--** [comment]
Would you like to open a PR so we can discuss more concretely on the code? :)
